ocaml/xapi/xapi_cluster_helpers.ml

```diff
@@ -59,8 +59,15 @@ let get_operation_error ~__context ~self ~op =
 
 let assert_operation_valid ~__context ~self ~op =
   begin match (Helpers.rolling_upgrade_in_progress ~__context), op with
-    | true, (`add | `remove | `destroy) -> raise Api_errors.(Server_error (not_supported_during_upgrade, []))
-    | _, _                              -> ()
+    | true, (`add | `remove | `destroy) ->
+      raise Api_errors.(Server_error (not_supported_during_upgrade, []))
+    | _, `remove -> ()
+    | _, _ -> begin
+      match Xapi_clustering.find_cluster_host ~__context ~host:Helpers.(get_localhost ~__context) with
```

Comment:
Shouldn't this check be done on the cluster host's `assert_operation_valid`? Otherwise you only ever look at the master here.

Reply:
The problem is that the cluster_host helpers only recognise the `enable` and `disable` enums

Reply:
Actually thats fine: we should be able to destroy a cluster host even if we never joined it.
Its just that the code in destroy will probably have to special case it, and if it was never joined, then do nothing on clusterd side, just delete the object from the DB.